service/bark/bark.go

```diff
@@ -0,0 +1,172 @@
+// if it's not already there. If the serverURL is empty, the DefaultServerURL is used. We're not validating the url here
+// on purpose, we leave that to the http client.
+func normalizeServerURL(serverURL string) string {
+	if serverURL == "" {
+		return DefaultServerURL
+	}
+
+	// Normalize the url
+	if !strings.HasPrefix(serverURL, "http") {
+		serverURL = "https://" + serverURL
+	}
+	if !strings.HasSuffix(serverURL, "/") {
+		serverURL = serverURL + "/"
+	}
+
+	return serverURL
+}
+
+// AddReceivers adds server URLs to the list of servers to use for sending messages. We call it Receivers and not
+// servers because strictly speaking, the server is still receiving the message, and additionally we're following the
+// naming convention of the other services.
+func (s *Service) AddReceivers(serverURLs ...string) {
+	for _, serverURL := range serverURLs {
+		serverURL = normalizeServerURL(serverURL)
+		s.serverURLs = append(s.serverURLs, serverURL)
+	}
+}
+
+// NewWithServers returns a new instance of Bark service. You can use this service to send messages to bark. You can
+// specify the servers to send the messages to. By default, the service will use the default server
+// (https://api.day.app/) if you don't specify any servers.
+func NewWithServers(deviceKey string, serverURLs ...string) *Service {
+	s := &Service{
+		deviceKey: deviceKey,
+		client:    defaultHTTPClient(),
+	}
+
+	if len(serverURLs) == 0 {
+		serverURLs = append(serverURLs, DefaultServerURL)
+	}
+
+	// Calling service.AddReceivers() instead of directly setting the serverURLs because we want to normalize the URLs.
+	s.AddReceivers(serverURLs...)
+
+	return s
+}
+
+// New returns a new instance of Bark service. You can use this service to send messages to bark. By default, the
+// service will use the default server (https://api.day.app/).
+func New(deviceKey string) *Service {
+	return NewWithServers(deviceKey)
+}
+
+// postData is the data to send to the bark server.
+type postData struct {
+	DeviceKey string `json:"device_key"`
+	Title     string `json:"title"`
+	Body      string `json:"body,omitempty"`
+	Badge     int    `json:"badge,omitempty"`
+	Sound     string `json:"sound,omitempty"`
+	Icon      string `json:"icon,omitempty"`
+	Group     string `json:"group,omitempty"`
+	URL       string `json:"pushURL,omitempty"`
+}
+
+func (s *Service) send(ctx context.Context, serverURL, subject, content string) (err error) {
+	if serverURL == "" {
+		return errors.New("server url is empty")
+	}
+
+	// Marshal the message to post
+	message := &postData{
+		DeviceKey: s.deviceKey,
+		Title:     subject,
+		Body:      content,
+		Sound:     "alarm.caf",
+	}
+
+	messageJSON, err := json.Marshal(message)
+	if err != nil {
+		return errors.Wrap(err, "marshal message")
+	}
+
+	u, err := url.Parse(serverURL)
+	if err != nil {
+		return errors.Wrap(err, "parse bark server url")
+	}
+	u.Path = path.Join(u.Path, "push")
+	pushURL := u.String()
```

Comment:
@moonD4rk I like your solution here, it's just a little redundant with what [http.NewRequestWithContext()](https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/request.go;l=869) does right after that. Internally it also uses `url.Parse()` so we get the URL validation automatically through that.

What do you think if we omit this part completely and just append `push` to the `serverURL`. Since `normalizeServerURL()` is always called when a new URL is added to the list of recipients, we can rely on the `serverURL` string ending with a `/`. And if the URL is generally invalid, `url.Parse()` in `http.NewRequestWithContext()` will return an error.

```suggestion
	pushURL := serverURL + "push"
```